Add first version of pytest infrastructure

davidrudlstorfer opened the pull request:
Includes first fixtures and tagging approach

davidrudlstorfer commented:
@isteinbrecher this is a work in progess - I've added the first pytest fixtures and also enabled the tagging approach. This turned out to be much more complicated than anticipated. 

I wanted that the "standard" tests would always execute - this approach was not doable with https://pypi.org/project/pytest_tagging/

Therefore, I simply implemented my own version which turns out to be very easy and doable

The tests can now be executed with:

`pytest` => execute tests which are not tagged
`pytest --4C` => execute tests which are tagged with fourc, i.e., `@pytest.mark.fourc`
`pytest --performance` => execute tests which are tagged with fourc, i.e., `@pytest.mark.performance`


It is also possible to add multiple tags to tests and also execute multiple tags at once

I am happy with the current status and I'll add the reference comparison function soon

davidrudlstorfer commented:
Also pytest coverage is added and works as expected - afterwards we can simply add 

1.  the coverage report of meshpy
2. the coverage badge to the readme

isteinbrecher commented on .github/actions/run_tests/action.yml:
```diff
@@ -49,5 +49,4 @@ runs:
         coverage html
         coverage report
         coverage-badge -o htmlcov/coverage.svg
```
We should be able to remove the `coverage run --rcfile=${{ inputs.coverage_config }} testing_main.py` call (and the corresponding calls to `coverage`, since `pytest` already picks up all "old" tests.

isteinbrecher commented on pyproject.toml:
```diff
@@ -39,6 +39,7 @@ dependencies = [
   "numpy-quaternion",
   "pre-commit",
   "pytest",
+  "pytest-cov",
```
Does not have to be decided in this PR, but do we want this in the default installation configuration? Or do we want a special "development" config?

davidrudlstorfer replied:
Good point - didn't think about that. For this PR I would leave it as is and we could adapt in within #152

isteinbrecher replied:
Sounds like a good plan!

isteinbrecher commented on pyproject.toml:
```diff
@@ -55,3 +56,11 @@ CI-CD = [
 ]
+
+[tool.pytest.ini_options]
+addopts = "-p pytest_cov --cov-report=term --cov-report=html --cov-fail-under=30 --cov=meshpy/ --cov=tutorial/ --cov-append"
```
Why only 30%? I think we can set that to 90. (With the exception of the performance tests).

Should we also add the `tests/` directory here? I personally think it makes sense to identify untested code in the testing framework.

davidrudlstorfer replied:
Locally I only execute the pytest framework and there I have 38% coverage - that's why I set it that low.

Normally the `tests/` is not added as only the coverage of the source code is of intereset

isteinbrecher commented on pyproject.toml:
```diff
@@ -55,3 +56,11 @@ CI-CD = [
+[tool.pytest.ini_options]
+addopts = "-p pytest_cov --cov-report=term --cov-report=html --cov-fail-under=30 --cov=meshpy/ --cov=tutorial/ --cov-append"
+markers = [
+  "fourc_arborx: tests in combination with 4C and ArborX",
```
Later on this is named `4C_ArborX` -> does this have to be the same string?

davidrudlstorfer replied:
i just differentiated the case style because I liked lowercase better within the decorators. And I thought within the CLI this approach looks better 😅 They do not have to be identical

isteinbrecher replied:
That makes sense. In the CLI I also prefer `4C`.

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+        "--4C_ArborX",
+        action="store_true",
+        default=False,
+        help="Execute 4C and ArborX based tests.",
```
According to the behaviour I saw, shouldn't this be:
```suggestion
        help="Execute basic tests and 4C and ArborX based tests.",
```

davidrudlstorfer replied:
Thanks - I changed that

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+
+    Currently configured options:
+        `pytest`: execute tests with no markers
+        `pytest --4C_ArborX`: execute tests with the `fourc_arborx` marker
```
If I read this code correctly, we always execute not tagged tests, right? Then this has to be adapted.

In general I really like the current approach (and kudos to the implementation!) of always executing the base tests and then allowing to selectively add test tags. Two comments from my side:
1. I would split 4C and ArborX, as a few people use MeshPy directly with 4C but don't have ArborX configured.
1. When running the performance tests, I would like to not run the untagged tests, is this possible?

davidrudlstorfer replied:
Good point - I'll split the 4C and ArborX tests

We could simply add a flag if the base tests are executed? Then this could be simply switched on and off for the performance tests as well all other tests

davidrudlstorfer replied:
Added an additional flag `--exclude-standard-tests` and it works as expected

`pytest` executes all standard tests
`pytest --4C` executes all standard tests + 4C tests
`pytest --4C --exclude-standard-tests` executes only the 4C tests

4C can be swapped with all other flags or an arbitrary number of flags can be combined, i.e.,

`pytest --4C --ArborX --CubitPy --Performance --exclude-standard-tests` executes only the special tests, i.e. 4C, ArborX, CubitPy and Performance test

Really nice solution in my opinion! Thanks for your input @isteinbrecher

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,137 @@
+
+
+@pytest.fixture(scope="function")
+def current_test_name(request: pytest.FixtureRequest) -> str:
```
Is it really necessary to provide the argument `requests` here? I am not sure if we have this available everywhere where we need to call this function. Maybe (for the beginning?) we should stick to the `env` variable approach. What do you think?

davidrudlstorfer replied:
I think https://stackoverflow.com/questions/17726954/py-test-how-to-get-the-current-tests-name-from-the-setup-method sums it up really nicely

This will also work for parameterized tests (we said we probably want those) and also it works for test functions as well as test methods.

The argument requests is always available once pytest is available.

Note: We do not need to provide the argument request. We simply call it via `current_test_name`.

isteinbrecher replied:
> Note: We do not need to provide the argument request. We simply call it via current_test_name.

Sorry, I was not aware of that, then this is perfectly fine the way it is.

isteinbrecher commented on tests/test_cosserat_curve.py:
```diff
@@ -50,30 +49,27 @@
     Cosserat curve."""
     coordinates = np.loadtxt(
-        os.path.join(testing_input, "test_cosserat_curve_centerline.txt"),
+        os.path.join(reference_file_directory, "test_cosserat_curve_centerline.txt"),
```
Do we want to passt this as a function argument or simply call the `reference_file_directory()` inside this function?

davidrudlstorfer replied:
As discussed in person we further use the pytest fixtures

isteinbrecher commented on tests/test_dummy.py:
```diff
@@ -0,0 +1,79 @@
+    print("reference_file_directory: ", reference_file_directory)
+
+    # approach to get temporary testing path
+    print(
+        "tmp_path: ", tmp_path
+    )  # pytest automatically keeps the last three runs for debugging and deletes everything that's older
```
Do you envision using this instead of the "testing-tmp" directory? In case CI tests fail we currently upload the `testing-tmp` directory to GitHub, to ease debugging. I would like to remain that possibility.

davidrudlstorfer replied:
I think the `tmp_path` approach is more viable as it is maintained through pytest and not our own implementation. 

For the CI we can also simply upload the test directory.

This also provides the benefit of even better management, e.g. see https://github.com/pytest-dev/pytest/issues/8141

We can then also only upload the data of the failed tests and trash the results of the successfull tests

isteinbrecher replied:
Sounds good!

isteinbrecher reviewed (commented):
@davidrudlstorfer thanks a lot for the effort! This looks really great and I really like it. I left some comments. Do you envision switching only a few modules in this PR or more? See my previous comment, we could simply merge the `pytest` config and already execute the "old" tests and the port file by file, but that is just a suggestion.

davidrudlstorfer commented:
@isteinbrecher thanks for your first round of review of the current status

In this PR I only wanted to update the current pytest (cosserat curve) and out of curiosity I've ported one additional feature (rotations). The remaining parts should be done in another PR

The only thing missing here is the implementation of the comparison function. As we've already briefly discussed I want to implement one function which automatically detects if either a string or file is handed over and then automatically compares the files or the string. I'll add that soon

isteinbrecher commented:
> The only thing missing here is the implementation of the comparison function. As we've already briefly discussed I want to implement one function which automatically detects if either a string or file is handed over and then automatically compares the files or the string. I'll add that soon

@davidrudlstorfer we already have that, the `compare_test_result` (which calls `compare_strings` under the hood) does exactly that. If I remember correctly, we were talking about unifying this for general text based files and vtk files (`compare_vtk`), although I am not sure if this does not introduce more complexity than simply having two functions for fundamental different file types.

davidrudlstorfer commented:
@isteinbrecher this would be ready for a review from your side.

In the last commit I've now added my suggestion for the new comparison feature. In my opinion this has a few advantages to the old version:

- one interface for testing comparisons (strings, vtk based objects, ... => easily extendable to other types, i.e. csv, ...)
- the interface is now clearer to the user => input are strings and nothing magical happens within the function (i.e., the retrieval of test names)
- Elaborate error retrieval and processing of errors: The actual differences are now printed for each failing test. This drastically improves the debugging rather than the manual file write out (which was not consistent across tests)

I've tested every path of the implementation and everything works accordingly!

I would be very happy if we could converge on a final pytest testing framework with the current suggested changes.

davidrudlstorfer commented:
This would now also enable to port everything to the new testing framework - as well as the deletion of the coverage configs and the test utils

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+
+    # compare strings for equality
+    if rtol is None and atol is None:
+        diff = list(
+            unified_diff(reference.splitlines(), result.splitlines(), lineterm="")
+        )
+        if diff:
+            print("Difference between reference and result:")
+            print("\n".join(list(diff)))
+            return False
+        else:
+            return True
```
As an example this now uses the standard python diff library to create a very nice diff between the strings rather than the old option to just write everything into a file and the user needs to manually compare it again.

davidrudlstorfer replied:
I've added back the option regarding the VSCode diff and extended it to even more usecases

I also added back the option to download the artifacts from Github

More on that in my summary later down below

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+    )
+
+    parser.addoption(
+        "--exclude-standard-tests",
```
Nice!

isteinbrecher commented on .github/actions/run_tests/action.yml:
```diff
@@ -46,8 +46,5 @@ runs:
-        coverage run --rcfile=${{ inputs.coverage_config }} -m pytest pytest_testing_cosserat_curve.py
-        coverage report
+        cd ..
+        pytest
```
Should we then set the testing flags via an `inputs:` variable? Something like `pytest_flags`. That could then simply be set in the workflows. This would make the `require_*` and the `coverage_config` flags obsolete.

Does it make sense to do this setup already in this PR? I think yes, but I will leave that up to you.

davidrudlstorfer replied:
One thing I have not yet considered is the following:

We have to adjust for the coverage in different pipeline runs. If the performance tests should not include the standard tests the coverage will be lower. Therefore, we somehow have to adjust the necessary coverage for differing runs.

In my opinion we could also just execute the standard tests in every config so the minimum coverage is reached.

davidrudlstorfer replied:
I've now adjusted the entire testing workflow on Github. Therefore, I removed all `require_*` flags and replaced them with the new pytest flags/markers

I've also simplified the performance tests and removed the separate workflow action

It should be noted that in the current state only the pytest tests are executed and the "old" manual testing framework is no longer executed

I would stay with this approach as in my opinion it is not straight forward to keep both testing strategies and we should move the old tests ASAP to the new approach once this is merged

davidrudlstorfer replied:
Additional Note: As discussed further above the base tests are always executed except for the performance tests => in this case only the performance tests are executed

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+            bool: true if comparison is successful, false otherwise
+        """
+
+        if type(reference) is not type(result):
```
This means we have to switch to `Path` objects to describe paths (at least in testing), correct? This is not used at all in the rest of MeshPy. Besides the type checking are there disadvantages of using a string here?

davidrudlstorfer replied:
Not really - in my opinion it would make sense to switch to path objects if we are at it. But let's discuss this in person


I'll just leave the first few search hints for "python string vs path" here.

https://medium.com/@guandika8/stop-representing-file-paths-in-python-using-strings-8312791f10f

https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/

https://www.btelligent.com/en/blog/best-practice-working-with-paths-in-python-part-2-1/

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+        """
+
+        if type(reference) is not type(result):
+            raise TypeError("Reference and result must be of the same type.")
```
We have a lot of tests, where we pass a string that is created during testing and compare it to a string saved in a reference file. This would mean that we have to first manually read the file and save it to a string variable and pass that to the function. That is likely just one or two lines of code, but they will be repeated a lot. Why not allow file vs. string comparisons?

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+    tolerance is given strings are compared for equality.
+
+    Args:
+        reference: The reference string.
+        result: The result string.
+        rtol: The relative tolerance.
+        atol: The absolute tolerance.
+        string_splitter: With which string the strings are split.
```
I really like the "new" docstring format when using type hints, that we don't have to specify the type here.

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+
+            if len(line_reference_splits) != len(line_result_splits):
+                raise AssertionError(
+                    f"Number of items in reference and result line differ!\n"
+                    + f"Reference line: {line_reference}\n"
+                    + f"Result line:    {line_result}"
```
Really nice, that is super helpful!

isteinbrecher commented on tests/test_cosserat_curve.py:
```diff
@@ -251,7 +262,9 @@ def test_cosserat_curve_mesh_warp_transform_boundary_conditions():
-    # Compare with the reference result
-    compare_test_result_pytest(
-        mesh.get_string(check_nox=False, header=False), rtol=1e-8, atol=1e-8
+    reference_string = get_string(
+        reference_file_directory / (current_test_name + ".dat")
     )
+    result_string = mesh.get_string(check_nox=False, header=False)
```
Like I mentioned before, do we really want to do this? We would have lines that look like this in almost every test case, producing a lot of duplicated code. Maybe, if we don't want the `check_nox=False, header=False` arguments all the time, we can consider allowing an `InputFile` to be passed to `compare_results`.

davidrudlstorfer replied:
I've entirely removed that logic, now we can even put in the mesh and everything is just one line.

More on that later down below in my summary.

isteinbrecher commented on tests/test_rotations.py:
```diff
@@ -0,0 +1,338 @@
+        0.25192421451158936,
+        0.4114279380770031,
+    ]
+    np.testing.assert_array_less(
```
I know this doesn't come from you, but do we want to switch all numpy array comparisons to `allclose` and provide a global standard for `rtol` and `atol`? Does not have to be part of this PR.

davidrudlstorfer replied:
Definitely - I did not even know that this option exists in numpy before. 

`allclose` makes more sense

davidrudlstorfer replied:
I removed a few occurences within this PR and would do the rest on a later occasion

isteinbrecher commented on tests/test_rotations.py:
```diff
@@ -0,0 +1,338 @@
+        np.testing.assert_array_less(
+            np.linalg.norm(t2 - t2_rot),
+            mpy.eps_quaternion,
+            "test_rotation_matrix: compare t2",
```
General comment, I used to do this in a lot of cases, write a short text of what is compared here. I stoped doing so a while ago, simply because I couldn't see any benefit of doing so. I would be fine with removing those (unless for some reason they are super meaningful) from the testing framework. Does not have to be part of this PR.

davidrudlstorfer replied:
I removed the comments which I was able to find quickly

isteinbrecher reviewed (commented):
@davidrudlstorfer thanks a lot for the work on the modernisation of the testing framework. I left some comments, maybe it makes sense to discuss them in person at some point. I would just like to get the functionality best fit for the type of testing done in MeshPy. Please don't get discourage by my comments, I really like this whole approach and greatly appreciate your effort!

Other question, if I run `pytest` it takes a couple of seconds to `gather` the tests. Do you have an idea how this could be speed up. This doesn't matter for CI, only for local testing.

davidrudlstorfer commented:
@isteinbrecher big thanks for your second round of review! Let's discuss the few open points in person to faster converge on one solution :)

> Other question, if I run pytest it takes a couple of seconds to gather the tests. Do you have an idea how this could be speed up. This doesn't matter for CI, only for local testing.

I think this is a combination of many things here. First of all it takes so long because `pytest` resolves all imports prior to running the tests. This takes some time in MeshPy because the whole module structure is not optimal. Once the imports are improved and modules are correctly split this should be faster (i.e. see #159)

https://stackoverflow.com/questions/16417546/how-to-speed-up-pytest

> Pytest imports all modules in the testpaths directories to look for tests. The import itself can be slow. This is the same startup time you'd experience if you ran those tests directly, however, since it imports all of the files it will be a lot longer. It's kind of a worst-case scenario.
> 
> This doesn't add time to the whole test run though, as it would need to import those files anyway to execute the tests.

The total test time should, therefore, stay the same in comparison to the current testing approach. The current testing suite takes around 3s on my machine and the current pytests take around 1s for collecting and running. I think once everything is ported we can look at the total execution time again. Also after restructuring the repo and improving the imports this should be even faster.

davidrudlstorfer commented on tests/test_cosserat_curve.py:
```diff
@@ -221,24 +225,31 @@ def test_cosserat_curve_mesh_warp():
 
 
-def test_cosserat_curve_mesh_warp_transform_boundary_conditions():
+def test_cosserat_curve_mesh_warp_transform_boundary_conditions(
+    reference_file_directory,
+    get_string,
+    current_test_name,
+    compare_results,
+):
```
As discussed in person we want to use the pytest approach

isteinbrecher commented on tests/performance_testing.py:
```diff
@@ -173,8 +172,7 @@ class TestPerformance(object):
 
     # Set expected test times.
-    expected_times = {}
-    expected_times["github-sisyphos-docker"] = {
```
Good decision to remove this hard coded host name!

Can you then also remove the one reference to this in the `README`?

```
(the reference time values and host name might have to be adapted in the file `<path_to_meshpy>/tests/performance_testing.py`)
```
replace with
```
(the reference time values might not suite your system)
```


Lets not put too much thought into that, as the README will likely be overhauled soon anyway.

davidrudlstorfer replied:
Thanks for the hint regarding the README. I totally forgot about that.

I've updated the readme with basic infos regarding the new testing framework - as you say this probably needs an update soon.

isteinbrecher commented on tests/test_cosserat_curve.py:
```diff
@@ -113,7 +107,7 @@ def test_cosserat_curve_translate_and_rotate():
         """Load the compare files and return a numpy array."""
         return np.loadtxt(
-            os.path.join(testing_input, f"{get_pytest_test_name()}_{name}.txt")
+            os.path.join(reference_file_directory, f"{current_test_name}_{name}.txt")
```
Could this maybe be the first case where we use the previously mentioned "get_reference_file_name" function?

davidrudlstorfer replied:
I've updated the currently ported tests to the new function which does everything automatically. More on that later down below

isteinbrecher commented on tests/test_cosserat_curve.py:
```diff
@@ -127,25 +121,32 @@ def load_compare(name):
-    compare_vtk_pytest(
-        os.path.join(testing_input, get_pytest_test_name() + ".vtu"),
-        result_name,
+    reference_path = Path(
+        os.path.join(reference_file_directory, current_test_name + ".vtu")
+    )
```
See above. And also later on ;)

Basically every time we use `current_test_name` at the moment.

isteinbrecher commented on tests/test_cosserat_curve.py:
```diff
@@ -127,25 +121,32 @@ def load_compare(name):
 
-def test_cosserat_curve_vtk_representation():
+def test_cosserat_curve_vtk_representation(
+    tmp_path, reference_file_directory, current_test_name, compare_results
```
The `tmp_path` is provided by pytest, correct? Can we make sure that the contents of this directory get uploaded to GitHub in case of test failure? Does not have to be part of this PR, just to mention it for future reference.

davidrudlstorfer replied:
Yes that's also added - more on that later down below (😅 sorry for the repetition)

isteinbrecher commented on tests/test_dummy.py:
```diff
@@ -0,0 +1,88 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+# -----------------------------------------------------------------------------
+"""Dummy test to demonstrate and ensure working pytest infrastructure."""
```
Will we delete this once all tests are ported?

davidrudlstorfer replied:
Yes we can delete that file - I've just created it so all the things are working properly and that the test selection works properly because no other tests are tagged yet.

This was also a little playground to see what is possible and that others (mainly you) can see how things should work out

isteinbrecher commented on tests/utils.py:
```diff
@@ -38,65 +38,6 @@
 import vtk
```
Do we still need this file once everything is ported? Would it make sense to move everything here to the `conftest.py` file?

davidrudlstorfer replied:
Yes - all the functionality is 100% ported and we can delete that file as well.

isteinbrecher reviewed (commented):
@davidrudlstorfer thanks for the changes, I am not sure if the PR was already ready for review, I sill left some comments 😅.

davidrudlstorfer commented:
@isteinbrecher sorry if it wasn't clear from my side: I am still working on the changes we discussed yesterday I hope to get them ready this evening or tomorrow morning

They will adress most of the points raised above, i.e., a very fancy way vastly simplify the result comparison with reference files

Currently its a real WIP - once the major things are resolved I'll push it 🚀

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+        # Get all set markers for current test (e.g. `fourc_arborx`, `cubitpy`, `performance`)
+        markers = [marker.name for marker in item.iter_markers()]
+
+        if config.getoption("--4C") and "fourc" in markers:
+            selected_tests.append(item)
+
+        if config.getoption("--ArborX") and "arborx" in markers:
+            selected_tests.append(item)
+
+        if config.getoption("--CubitPy") and "cubitpy" in markers:
+            selected_tests.append(item)
+
+        if config.getoption("--performance-tests") and "performance" in markers:
+            selected_tests.append(item)
```
Really nice suggestion 🙏🏻 applied it in the latest commit

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+        Path: A Path object representing the full path to the reference file directory.
+    """
+
+    testing_path = os.path.abspath(os.path.dirname(__file__))
+    return Path(testing_path) / "reference-files"
```
Nice thanks for improving the Paths :)

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+    if not compare[0]:
+        raise AssertionError("\n".join(compare[1]))
+
+    return compare[0]
```
This is added within this PR and I will write more on that later down below

isteinbrecher commented:
I am sorry, I didn’t have a close enough look before reviewing 😅

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,350 @@
+    def _compare(
+        reference: Union[Path, str],
+        result: Union[Path, str],
+        rtol: float = 1e-14,
+        atol: float = 1e-14,
```
Very good catch!

I hope I fixed it correctly

davidrudlstorfer commented:
@isteinbrecher First of all - sorry that this took so long from my side - somehow I wanted to make it really fancy and with the current status of this PR at least my own expectations are met 😅 Even today all the changes took much longer than anticipated to include all the functionality as universally applicable as possible.

I am very curious about your opinion 🫣

1. All points we discussed yesterday in person are now adressed
2. I've addressed all conversations above

I'll quickly summarize the most important features of this PR

1. Tagging/Marking
We can now tag/mark any test we want like
```python
@pytest.mark.fourc
test_4C():
    pass
```
and execute them with
```bash
pytest --4C
```

2. I provided the following easy to use fixtures
    - `reference_file_directory` => returns the path to the reference file directory
    - `current_test_name` => returns the string of the current test name
    - `get_string` => returns the string of any given file
    - `get_corresponding_reference_file_path` => get the path to the corresponding dat file name of the current test
    - `compare_results` => this is the most important and somehow my holy function. With this you can compare anything we (currently) need, i.e., strings + strings, file + string, vtk+vtk, **file + input_file**
    This results now in these simple asserts for each test case:
    ```python
        assert compare_results(
        get_corresponding_reference_file_path, mesh, rtol=1e-8, atol=1e-8
    )
    ```
    Nothing more is needed! `get_corresponding_reference_file_path` automatically provides the path to the identically named .dat file and the string is automatically read from the mesh. In my opinion this is now an easily understandable code part which also does not need many lines.
    
 3. I've switched the testing framework within the github actions to pytest
 
 4. I've updated the README (with very little effort 😅)
 
 5. If a test fails it now looks like this. For each test the following output is written to the terminal (locally and within the pipeline):
 ```bash
 ----------------------------- Captured stdout call -----------------------------
========= Test: 'test_cosserat_curve_mesh_warp_transform_boundary_conditions' =========
String comparison with tolerance failed!
Numbers do not match within given tolerance!
Reference line: MAT 1 MAT_BeamReissnerElastHyper YOUNG -1.0 POISSONRATIO 0.5 DENS 0.0 CROSSAREA 0.007853981633974483 SHEARCORR 1 MOMINPOL 9.817477042468105e-06 MOMIN2 4.9087385212340526e-06 MOMIN3 4.9087385212340526e-06
Result line:    MAT 1 MAT_BeamReissnerElastHyper YOUNG -1.0 POISSONRATIO 0.0 DENS 0.0 CROSSAREA 0.007853981633974483 SHEARCORR 1 MOMINPOL 9.817477042468105e-06 MOMIN2 4.9087385212340526e-06 MOMIN3 4.9087385212340526e-06
Result string saved to: '/tmp/pytest-of-runner/pytest-0/test_cosserat_curve_mesh_warp_0/test_cosserat_curve_mesh_warp_transform_boundary_conditions_result.txt'.
```
Further, within the local IDE (VSCode) the same diff is opened as before. And on Github the temporary testing directory is uploaded (and the new result files can be downloaded) (for example see this run https://github.com/imcs-compsim/meshpy/actions/runs/12677922314/job/35334350367)

6. If I did not forget anything the only thing we have to adjust in the future is the necessary coverage once we have added more tests. We also have to differentiate between the testing machines and their coverage (i.e., the performance test run only executes the performance tests and not the standard tests => therefore, the coverage is lower)

7. And once everything is added we can also delete the old test utils, the dummy test and the dummy test reference file

Concluding - I think this now provides a good base for the complete switch to pytest. Thanks for the time you invested in all the review rounds and into our in-person meeting 🙏🏻

I'm really curious what you have to say 😄

davidrudlstorfer commented:
I've ported one additional small file and the porting works really really nice

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,512 @@
+        result: Union[Path, str, InputFile],
+        rtol: Optional[float] = None,
+        atol: Optional[float] = None,
+        input_file_string_check_nox: bool = False,
+        input_file_string_check_header: bool = False,
```
These two parameters are very specific, I am not sure if they make sense here. Maybe we can control the check via the `mpy` object. Alternatively we could "bundle" those arguments in a dictionary here, e.g.,
```python
    def _compare(
        reference: Union[Path, str],
        result: Union[Path, str, InputFile],
        rtol: Optional[float] = None,
        atol: Optional[float] = None,
        input_file_kwargs: dir={
            "input_file_string_check_nox": False,
            "input_file_string_check_header": False,
       },
        **kwargs,
    )
```
and then pass `input_file_kwargs` to `result.get_string(**input_file_kwargs)`.

davidrudlstorfer replied:
I really like the approach with the bundled arguments in a dictionary - I've added it in the latest commit

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,512 @@
+            try:
+                return compare_strings(reference_string, result, rtol, atol, **kwargs)
+            except AssertionError as error:
+                return handle_unequal_strings(
+                    error, tmp_path, current_test_name, result, reference
+                )
+
+        # Compare a file with an InputFile
+        elif isinstance(reference, Path) and isinstance(result, InputFile):
+            reference_string = get_string(reference)
+            result_string = result.get_string(
+                check_nox=input_file_string_check_nox,
+                header=input_file_string_check_header,
+            )
+
+            try:
+                return compare_strings(
+                    reference_string, result_string, rtol, atol, **kwargs
+                )
+            except AssertionError as error:
+                return handle_unequal_strings(
+                    error, tmp_path, current_test_name, result_string, reference
+                )
+
+        # Compare two universal files
+        elif isinstance(reference, Path) and isinstance(result, Path):
+            if reference.suffix != result.suffix:
+                raise RuntimeError(
+                    "Reference and result file must be of same file type!"
+                )
+            elif reference.suffix in [".vtk", ".vtu"]:
+                return compare_vtk_files(reference, result, rtol, atol)
+            else:
+                raise NotImplementedError(
+                    f"Comparison is not yet implemented for {reference.suffix} files."
+                )
+        else:
+            raise TypeError("Reference and result must be either string or Path.")
+
+    return _compare
```
I really like this approach of having one clearly structured function to do all the comparisons!

One thought, in most cases we end up actually comparing strings to each other. Should we consider a slightly different approach that will allow for an easier extension to other types of input (see my previous comment regarding the comparison between two meshes).

Something like (quickly copied together pseudo code)
```python
if ref_is_file and res_is_file:
  # Here we implement true file comparisons
  if ref_type is "vtk/vtu" and res_type is "vtu/vtk:
     return compare_vtk_files(reference, result, rtol, atol)
  else:
     raise error("Comparison for the file types "xxx" is not implemented")
else:
    # Here we have string comparisons
    def get_string2(data):
        """Could also be done with the previously defined get_string but also locally here"""
        if data_type is str:
            return data
        elif data_type is InputFile:
            return data.get_string(**some_args)
        else:
            return get_string(data)

     reference_string, result_string = [get_string2(data) for data in [reference, result])
     try:
          return compare_strings(
          reference_string, result_string, rtol, atol, **kwargs
    )
     except AssertionError as error:
          # Do we want to always call this function to improve the console output, or just in case that reference is a file?
          return handle_unequal_strings(
               error, tmp_path, current_test_name, result_string, reference
          )
        
```

davidrudlstorfer replied:
Thanks for the nice suggestion - this has also minimized duplicated code. I've added the new version in the last commit

Regarding the question in the code - we only use the handle unequal strings function if the reference is a file, because otherwise we could not make a diff in Vscode etc.

In the string-string comparisons the same error is printed to the CLI with the Assertions which are executed - so the output is more or less the same - except for the diff

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,512 @@
+    return reader.GetOutput()
+
+
+def handle_unequal_strings(
```
Nice that this functionality is retained! Is this also the function that is responsible that the "failed" results get written to disc and then uploaded to GitHub?

davidrudlstorfer replied:
Yes - exactly. The files are written to the tmp directory and the upload just uses the temp directory in the Github Actions

isteinbrecher reviewed (commented):
Thanks for the effort @davidrudlstorfer! I really like the result. I left some comments.

davidrudlstorfer commented on .github/workflows/testing.yml:
```diff
@@ -47,16 +47,12 @@ jobs:
         with:
           name: ${{github.job}}-${{ matrix.os-version }}-python${{ matrix.python-version }}-${{github.run_number}}
-          path: ${{github.workspace}}/tests/testing-tmp/
+          path: /tmp/pytest-of-runner/
```
Good catch :eyes: 

Yesterday I read some docs that the path is identical for all Github hosted runners but as one can see here https://github.com/imcs-compsim/meshpy/actions/runs/12677922314 the path actually differs on Mac and Windows

I'll fix it

davidrudlstorfer replied:
Now the upload works for all systems - in the pipeline i simply use the provided tmp directory by Github and upload it in the final step

See https://github.com/imcs-compsim/meshpy/actions/runs/12690733332 for all uploaded artifacts

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,512 @@
+
+        return corresponding_reference_file_path
+
+    return _get_corresponding_reference_file_path()
```
I've added the old logic to enable editing the corresponding reference file path

davidrudlstorfer commented:
Of course, it took much longer than expected - it felt like 50 pipeline runs :smiling_face_with_tear: 

Now everything works and the artifact upload works for all containers (macOS, Windows, ubuntu, 4C container)

I've adressed all of your comments and this would be ready for your next review @isteinbrecher

davidrudlstorfer commented:
This would be a failed pipeline run where you can download the artifact from each job

https://github.com/imcs-compsim/meshpy/actions/runs/12692452173

isteinbrecher commented:
@davidrudlstorfer thanks for the work, I think we are very close to merging this. I took the liberty and added two commits to the branch, as a suggestion:
```
b769dca Allow for general types in compare function
```
This commit allows for comparison of all possible combinations of `Path`, `str`, and `InputFile` with each other.

```
9766fde Refactor assertions and return values from compare functions
```
While adding the changes form the previous commit, I realised that I missed the changes in the `compare_strings_with_tolerance` function which raise an error. This should not be done in this function, as we use this one in some other projects where we rely that the return value is either True or False. I changed that accordingly.

While looking at this, I saw that we had some (at least to me) not so intuitive behaviours in the code: We had functions that return True or raise an error, but never return False. At least to me that seemed weird. I tried to unify this, that the compare functions we actually will call from within the testing framework now either run without complaining or raise an assertion. This also allows to simply call the renamed compare function in the testing with
```python
assert_results_equal(...
```
Without the leading `assert` - I personally like that, since this eliminates the possible error that someone calls the compare function and forgets the `assert`.

I am curios about your opinion on these changes.

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+
+
+@pytest.fixture(scope="function")
+def assert_results_equal(get_string, tmp_path, current_test_name) -> Callable:
```
That's a good idea regarding the name change! Now it better reflects what it does and it makes sense to no longer return true or false.

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+        Function to compare results.
+    """
+
+    def _compare(
```
Sorry that I also made the error - here it would be nice if the inner function is named similarly/identical to the outer one, i.e.,

`_assert_results_equal`

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+        rtol: Optional[float] = None,
+        atol: Optional[float] = None,
+        input_file_kwargs: dict = {
+            "check_nox": False,
+            "header": False,
```
That's a really nice idea! Improves the readability by a lot

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+            "header": False,
+        },
+        **kwargs,
+    ):
```
```suggestion
    ) -> None:
```

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+        # String based comparison
+        else:
+
+            def get_string_to_compare(data):
+                """Get the string depending on the data type that will actually
+                be compared."""
+                if isinstance(data, str):
+                    return data
+                elif isinstance(data, Path):
+                    return get_string(data)
+                elif isinstance(data, InputFile):
+                    return data.get_string(**input_file_kwargs)
+                else:
+                    raise TypeError(f"Data type {type(data)} not implemented.")
+
+            # retrieve strings to compare
+            [reference_string, result_string] = [
+                get_string_to_compare(data) for data in [reference, result]
+            ]
```
Currently I do not like that construct that much. First of all we could completely remove the inner by simply generalizing the `get_string` function. For me the main problem is that on the bottom we retrieve the strings but the input could be a string as well. So we have function calls to convert a string into a string. This entire construct extremely increases the complexity of what is actually happening and I currently see no added benefit.

If you like we can keep it for sure as this is only my opinion

davidrudlstorfer replied:
I am sold - let's keep the now implemented more general way. Big thanks for going the extra mile and improving everything - especially by expanding the `get_string` function

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+        # String based comparison
+        else:
+
+            def get_string_to_compare(data):
+                """Get the string depending on the data type that will actually
```
Here we should also add the type hinting and the current style of docstring with args and return if we decide on keeping it.

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+            compare, error_message = compare_strings(
+                reference_string, result_string, rtol, atol, **kwargs
+            )
+            if not compare:
+                if isinstance(reference, Path):
```
With the try, except and split into the pure string comparison this nested if construction was not necessary (in the beginning I had the same code but I removed it 😅)

isteinbrecher replied:
Yes, lets go back to your approach in this function, see my comment below.

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+    rtol: Optional[float] = None,
+    atol: Optional[float] = None,
+    string_splitter: str = " ",
+) -> Tuple[bool, str]:
```
This weird return type is what I wanted to prevent - let's just raise an assertion error and catch it above? Then we do not need to return anything in this function as well and can simply reuse it

isteinbrecher replied:
Yes, lets go back to your approach in this function, see my comment below.

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+    else:
+        return compare_strings_with_tolerance(
+            reference, result, rtol, atol, string_splitter=string_splitter, output=True
+        )  # type: ignore
```
Why do we need this ignore?

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+        comparisons.
+    """
+
+    def get_return_values(flag, message):
+        """Get the data structure that shall be returned from this function."""
+        if output:
+            return flag, message
+        else:
+            return flag
```
This entire construct could also be simply removed if we just raise the Assertion Error and catch it above - for me that's a lot of added complexity

isteinbrecher replied:
See my previous comment

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+    *,
+    string_splitter=" ",
+    output=False,
+) -> Union[bool, Tuple[bool, str]]:
```
I think the function should have a fixed return value that would simplify a lot.

isteinbrecher replied:
Maybe I can give a little context here. I view this function as a utility that allows to compare strings, where floating point values (or integers) can be compared using numerical tolerances. It is also used in some other projects, e.g., CubitPy. I would like to not throw anything in this function (unless an actually error occurs), because in my opinion the user of this function is responsible to handle the result of the comparison. Now I really like your additional error messages, so I choose an approach very similar to the one in `compare_grids` for `vtk` comparisons (on that note, maybe we also want to retrieve the error message from `compare_grids` and print it for failed tests). If I remember correctly, I saw this approach with the optional information in the return value in some `numpy` functions.

This might not be the best way, but for the pure compare functions I would like to keep that approach (unless there is some nice alternative without assertions).

We can basically decide how to handle the rest in MeshPy and there I am completely open. I am fine with going back to your previous approach with the assertions, then we basically just have to check the return value from `compare_strings_with_tolerance` inside `compare_strings` and throw the respective assertion. Then everything else could go back to the way you had it (maybe without the `return True` in the functions to simply keep them `void` and only throw an assertion in case of a failed comparison).

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+    current_test_name: str,
+    result: str,
+    reference_path: Path,
+):
```
```suggestion
) -> None:
```

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,537 @@
+            ]
+
+            # compare strings and handle non-matching strings
+            compare, error_message = compare_strings(
+                reference_string, result_string, rtol, atol, **kwargs
```
That's also one point I explicitly wanted to prevent. In my opinion it's not really nice to return a tuple only in the case if it fails. If it succeeds it just returns an empty string which is completely unused.

My thought behind catching the assertion was that the function can also be used entirely on it's own. Now it's only usable in combination with the code down below.

isteinbrecher replied:
Yes, lets go back to your approach in this function, see my comment below.

davidrudlstorfer reviewed (commented):
@isteinbrecher thanks a lot for extending the functions!

Unfortunately, I do not fully agree with you on all changes. In many cases a lot of complexity and nested functions were added where it is not necessary at all. Also functions should always return one thing and not just empty strings depending on the outcome.

We could probably increase the simplicity of the code by a lot. I also fully agree with you if you decide to merge the current changes.

davidrudlstorfer commented:
If you like I am more than happy to also provide a suggestion on my comments from above - I think in the end it is worth it if we provide a really nice testing framework 🙏🏻

isteinbrecher reviewed (commented):
Thanks for your comments @davidrudlstorfer. I left some explanations. I will add another commit trying to revert my changes to address most of your major points. We can use that as a starting point for further discussion or simply revert it again.

isteinbrecher commented on tests/conftest.py:
```diff
@@ -0,0 +1,519 @@
+    """
+
+    print(f"========= Test: '{current_test_name}' =========")
+    print(error)
```
Do we really want to print the error here? Shouldn't that be done by `pytest` itself?

davidrudlstorfer replied:
Yes - we can definetily remove these two lines - I'll remove it

davidrudlstorfer commented:
@isteinbrecher thanks for adding the changes and explanations!

If I understand you correctly your major concern is that you want to use the function externally as well. What do you think if we just provide a wrapper for the external use?

I added this in the last commit - this function now does exactly the same thing you implemented but is just a wrapper around the standard `compare_strings_with_tolerance` function.

This way the function signature is identical within meshpy for `compare_strings_equality` and `compare_strings_with_tolerance`.

I am open for a better name of the function which also returns stuff 😅

Curious about your opinion

davidrudlstorfer commented on tests/conftest.py:
```diff
@@ -0,0 +1,534 @@
+
+            # compare strings and handle non-matching strings
+            try:
+                return compare_strings(
```
Done - good catch

davidrudlstorfer commented:
Perfect 🤩 I've addressed the 2 minor remarks. I'll squash everything into a few manageable commits and then force push it

isteinbrecher reviewed (approved):
Thanks again for the effort, lets get this merged!
